plugin/phpactor.vim

```diff
@@ -145,24 +145,50 @@ endfunction
 """"""""""""""""""""""""
 " Extract method
 """"""""""""""""""""""""
-function! phpactor#ExtractMethod()
-    let selectionStart = phpactor#_selectionStart()
-    let selectionEnd = phpactor#_selectionEnd()
 
-    call phpactor#rpc("extract_method", { "path": phpactor#_path(), "offset_start": selectionStart, "offset_end": selectionEnd, "source": phpactor#_source()})
+function! phpactor#ExtractMethod(...)
+    let positions = {}
+
+    if 0 == a:0 " Visual mode - backward compatibility
+        let positions.start = phpactor#_selectionStart()
+        let positions.end = phpactor#_selectionEnd()
+    elseif a:1 ==? 'v' " Visual mode
+        let positions.start = phpactor#_selectionStart()
+        let positions.end = phpactor#_selectionEnd()
+    else " Linewise or characterwise motion
+        let linewise = 'line' == a:1
+
+        let positions.start = s:getStartOffsetFromMark("'[", linewise)
+        let positions.end = s:getEndOffsetFromMark("']", linewise)
+    endif
+
+    call phpactor#rpc("extract_method", { "path": phpactor#_path(), "offset_start": positions.start, "offset_end": positions.end, "source": phpactor#_source()})
 endfunction
 
-function! phpactor#ExtractExpression(isSelection)
+function! phpactor#ExtractExpression(type)
+    let positions = {}
 
-    if a:isSelection 
-        let selectionStart = phpactor#_selectionStart()
-        let selectionEnd = phpactor#_selectionEnd()
-    else
-        let selectionStart = phpactor#_offset()
-        let selectionEnd = v:null
+    if v:true == a:type  " Invoked from Visual mode - backward compatibility
+        let positions.start = phpactor#_selectionStart()
+        let positions.end = phpactor#_selectionEnd()
+    elseif v:false == a:type " Invoked from an offset - backward compatibility
+        let positions.start = phpactor#_offset()
+        let positions.end = v:null
+    elseif a:type ==? 'v' " Visual mode
+        let positions.start = phpactor#_selectionStart()
+        let positions.end = phpactor#_selectionEnd()
+    else " Linewise or characterwise motion
+        let linewise = 'line' == a:type
+
+        let positions.start = s:getStartOffsetFromMark("'[", linewise)
+        let positions.end = s:getEndOffsetFromMark("']", linewise)
```

Comment:
So `v` and `True` are the same? What does linewise / characterwise do? (not familiar with this)

Reply:
I guess the difference is that expression can work with just a single offset, not sure if method works like that - in theory it could extract the expression under the cursor.

Reply:
it's a vim notion so I will refer you to the doc, since it's been a while I don't remember exactly and it will be more accurate this way:
```vim
:h linewise
```